### PR TITLE
stats: do not wait for data unchanged when auto analyze

### PR DESCRIPTION
**Pull request by lamxTyler**
## What have you changed? (mandatory)

Before this PR, if the table is continous updated, we cannot auto analyze the table even if `modify count / count` is greater than auto-analyze ratio, because we have the limit that the table must not be modified within a time peroid. This PR removes the restriction.

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

Unit test

PTAL @coocood @winoros @zz-jason @fipped

### CODE REVIEW

**Comment by winoros** [comment]
What will happen if the following scene occur:
The limit is x. The continuous writing row count is more the 2x of it.

Will it trigger multiple analyze in a row?

File: statistics/update.go [hunk 1]
```diff
@@ -533,28 +533,37 @@ const (
 // AutoAnalyzeMinCnt means if the count of table is less than this value, we needn't do auto analyze.
 var AutoAnalyzeMinCnt int64 = 1000
 
+func tableAnalyzed(tbl *Table) bool {
```
Comment (shenli) [inline statistics/update.go]: Need comments.

**Comment by lamxTyler** [comment]
@winoros  Yes.

File: statistics/update.go [hunk 2]
```diff
@@ -533,28 +533,38 @@ const (
...
 
+// tableAnalyzed checks if the table is analyzed.
+func tableAnalyzed(tbl *Table) bool {
+	for _, col := range tbl.Columns {
+		if col.Count > 0 {
+			return true
+		}
+	}
+	for _, idx := range tbl.Indices {
+		if idx.Len() > 0 {
```
Comment (shenli) [inline statistics/update.go]: What is the `idx.Len()`? The length of histogram?

File: statistics/update.go [hunk 3]
```diff
@@ -533,28 +533,35 @@ const (
...
+}
+
+// needAnalyzeTable checks if we need to analyze the table. If the table is not analyzed,
+// we need to analyze it when it has not been modified for a time. If the table is analyzed,
+// we need to analyze it when the `modify count / count` is greater than autoAnalyzeRatio.
+func needAnalyzeTable(tbl *Table, limit time.Duration, autoAnalyzeRatio float64) bool {
+	analyzed := tableAnalyzed(tbl)
+	if !analyzed {
+		t := time.Unix(0, oracle.ExtractPhysical(tbl.Version)*int64(time.Millisecond))
+		return time.Since(t) >= limit
```
Comment (coocood) [inline statistics/update.go]: The `AutoAnalyzeMinCnt` is not considered?

Comment (zimulala) [reply statistics/update.go]: It's considered in line 597.

File: statistics/update.go [hunk 4]
```diff
@@ -536,35 +536,32 @@ var AutoAnalyzeMinCnt int64 = 1000
...
 	for _, idx := range tbl.Indices {
-		if idx.Len() > 0 {
+		if idx.Histogram.Len() > 0 {
 			return true
 		}
 	}
 	return false
 }
 
+// needAnalyzeTable checks if we need to analyze the table. If the table is not analyzed,
```
Comment (shenli) [inline statistics/update.go]: If the table has never been analyzed,

File: statistics/update.go [hunk 5]
```diff
@@ -536,35 +536,32 @@ var AutoAnalyzeMinCnt int64 = 1000
...
+// we need to analyze it when it has not been modified for a time. If the table is analyzed,
```
Comment (shenli) [inline statistics/update.go]: If the table had been analyzed before

**Comment by lamxTyler** [comment]
PTAL @coocood @shenli

**Comment by shenli** [comment]
LGTM

File: statistics/update.go [hunk 6]
```diff
@@ -533,28 +533,35 @@ const (
...
+			return true
 		}
 	}
-	return true
+	return false
+}
+
+// needAnalyzeTable checks if we need to analyze the table. If the table has never been analyzed,
+// we need to analyze it when it has not been modified for a time. If the table had been analyzed before,
+// we need to analyze it when the `modify count / count` is greater than autoAnalyzeRatio.
```
Comment (zz-jason) [inline statistics/update.go]: how about this:
```go
// needAnalyzeTable checks if we need to analyze the table.
// 1. If the table has never been analyzed, we need to analyze it when it has
//    not been modified for a time.
// 2. If the table had been analyzed before, we need to analyze it when
//    "tbl.ModifyCount/tbl.Count > autoAnalyzeRatio".
```

**Pull request merged**